### PR TITLE
DM-7028: Port to Python 3

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

File: tests/dateTime.py [hunk 1]
```diff
@@ -1,29 +1,31 @@
...
 # This program is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
-# 
-# You should have received a copy of the LSST License Statement and 
-# the GNU General Public License along with this program.  If not, 
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
```
Comment (PaulPrice) [inline tests/dateTime.py]: Commit also includes PEP8 cleanup.

Comment (timj) [inline tests/dateTime.py]: I hoped you were going to let me get away with that one. My editor always strips spaces and I didn't catch this whitespace cleanup.

File: tests/dateTime.py [hunk 2]
```diff
@@ -54,74 +56,74 @@ def testLeapSecond(self):
...
+        ts = DateTime(long(1192755506000000000))
+        self.assertEqual(ts.nsecs(DateTime.UTC), long(1192755473000000000))
+        self.assertEqual(ts.nsecs(DateTime.TAI), long(1192755506000000000))
+        self.assertEqual(ts.nsecs(), long(1192755506000000000))
         self.assertAlmostEqual(ts.get(DateTime.MJD, DateTime.UTC), 54392.040196759262)
 
     def testNow(self):
         successes = 0
-        for i in xrange(10):       # pylint: disable-msg=W0612
+        for i in range(10):       # pylint: disable-msg=W0612
```
Comment (PaulPrice) [inline tests/dateTime.py]: What's the `pylint` comment suppress?

File: tests/dateTime.py [hunk 3]
```diff
@@ -54,74 +56,74 @@ def testLeapSecond(self):
...
             secs = time.time()
             ts = DateTime.now()
-            diff = ts.nsecs(DateTime.UTC) / 1.0e9 - secs 
+            diff = ts.nsecs(DateTime.UTC) / 1.0e9 - secs
```
Comment (PaulPrice) [inline tests/dateTime.py]: Since you touched the line, may as well clean it up: no spaces around `/`.

File: tests/dateTime.py (hunk 2 above)
Comment (timj) [inline tests/dateTime.py]: No idea. Added by @ktlim back in commit 9a8ab201bc5b1828631fec02213a4460b6a9ee6b

Comment (timj) [inline tests/dateTime.py]: Ah. It's unused variable warning. `i` is not actually used in the loop.

File: tests/dateTime.py (hunk 3 above)
Comment (timj) [inline tests/dateTime.py]: The price one pays for removing a trailing space...

File: tests/dateTime.py (hunk 1 above)
Comment (PaulPrice) [inline tests/dateTime.py]: I think RHL chastised me about that kind of editor setting several years ago.

File: tests/dateTime.py (hunk 2 above)
Comment (PaulPrice) [inline tests/dateTime.py]: I think tradition is to call such a variable `_`.

File: tests/dateTime.py (hunk 3 above)
Comment (PaulPrice) [inline tests/dateTime.py]: ~(;

File: python/lsst/daf/base/baseLib.i [hunk 4]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+    On Python 3 all ints are LongLong but we need to be able to store them
+    in Int containers if that is what is being used (testing for truncation).
+    Python3 uses sys.maxsize as an equivalence for the cross over from int to long.
+
+    If there is no pre-existing value we have to decide what to do. For now
+    we pick Int if the value is less than maxsize.
+
+    On Legacy Python always returns None.
+
+    Returns None if the value supplied is not an integral value."""
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Closing `"""` should be on a line by itself.

File: python/lsst/daf/base/baseLib.i [hunk 5]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+    is assumed to be a scalar.
+
+    On Python 3 all ints are LongLong but we need to be able to store them
+    in Int containers if that is what is being used (testing for truncation).
+    Python3 uses sys.maxsize as an equivalence for the cross over from int to long.
+
+    If there is no pre-existing value we have to decide what to do. For now
+    we pick Int if the value is less than maxsize.
+
+    On Legacy Python always returns None.
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Sentence doesn't parse.

File: python/lsst/daf/base/baseLib.i [hunk 6]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+
+    Returns None if the value supplied is not an integral value."""
+    useType = None
+
+    # Should we do this? It will possibly lead to different behavior
+    # between Python versions where passing in long(3) in Legacy Python
+    # will be different to passing in "long(3)" on Python 3.
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: I think I lean towards having consistency across python versions.

File: python/lsst/daf/base/baseLib.i (hunk 5 above)
Comment (timj) [inline python/lsst/daf/base/baseLib.i]: Would you prefer "On Python 2 always returns None"?

File: python/lsst/daf/base/baseLib.i [hunk 7]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+        return useType
+
+    if isinstance(value, numbers.Integral):
+        try:
+            containerType = _propertyContainerElementTypeName(container, name)
+            if containerType == "Int":
+                if value <= sys.maxsize:
+                    useType = "Int"
+            elif containerType == "LongLong":
+                useType = "LongLong"
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Put the `if.. elif` block in the `try`'s `else` block?  You're not capturing an exception from anything except the `containerType` line.

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: ok.

File: python/lsst/daf/base/baseLib.i (hunk 5 above)
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Oh, I see.  "On Legacy Python, always returns None."  (comma).  Or "Always returns None for Legacy Python."

File: python/lsst/daf/base/baseLib.i [hunk 8]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+        except lsst.pex.exceptions.NotFoundError:
+            # nothing in the container so choose based on size. Safe option is to
+            # always use LongLong
+            if value <= sys.maxsize:
+                useType = "Int"
+            else:
+                useType = "LongLong"
+    return useType
+
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Need an extra line here?

File: python/lsst/daf/base/baseLib.i [hunk 9]
```diff
@@ -219,7 +271,7 @@ def _PS_toDict(self):
     d = {}
     for name in self.names():
         v = self.get(name)
-    
+
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Whitespace cleanup.

File: tests/testExecutables.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import unittest
+import lsst.utils.tests
+
+
+class UtilsBinaryTester(lsst.utils.tests.ExecutablesTestCase):
+    pass
+
+EXECUTABLES = None
+UtilsBinaryTester.create_executable_tests(__file__, EXECUTABLES)
```
Comment (PaulPrice) [inline tests/testExecutables.py]: Why is this here?  After a bit of digging, it seems it needs to be present to get non-`unittest` tests working with our new test system, but the interface is strange.  For example, why is `EXECUTABLES = None`?  That gives the impression that this is a placeholder or somesuch, but I don't think that's right.

Comment (timj) [inline tests/testExecutables.py]: This was all dealt with in the SQR-012 work. If `EXECUTABLES` is `None` the system discovers the executables on its own. If you give a list it will just run those executables. Are you saying it should have been using an optional keyword arg, something like:
```python
create_executable_tests(__file__, exes=EXECUTABLES)
```
?

File: tests/PropertySet_2.py [hunk 11]
```diff
@@ -357,10 +357,12 @@ def testToDict(self):
 
         self.assertIsInstance(d2["bool"], bool)
         self.assertEqual(d2["bool"], True)
+        # Newer versions of SWIG can return int here
```
Comment (PaulPrice) [inline tests/PropertySet_2.py]: Just a note?  Not doing anything about it for now?

File: tests/PropertySet_2.py [hunk 12]
```diff
@@ -357,13 +357,11 @@ def testToDict(self):
 
         self.assertIsInstance(d2["bool"], bool)
         self.assertEqual(d2["bool"], True)
-        # Newer versions of SWIG can return int here
-        self.assertIsInstance(d2["short"], long)
+        self.assertIsInstance(d2["short"], int)
```
Comment (PaulPrice) [inline tests/PropertySet_2.py]: Oh, you are doing something about it.  Squash into previous.

File: tests/PropertySet_2.py (hunk 11 above)
Comment (timj) [inline tests/PropertySet_2.py]: I think there is a later commit to remove this line now that SWIG has been updated and we can't go back. @ktlim asked for a note in the code there so we would understand why the test fails. In modern SWIG they come back as int. He did not want a generic test for `number.Integral`.

File: tests/PropertySet_2.py (hunk 12 above)
Comment (timj) [inline tests/PropertySet_2.py]: Yes, driven by the SWIG update (when the code was first written we did not realize that we had to update SWIG).

File: tests/testExecutables.py (hunk 10 above)
Comment (PaulPrice) [inline tests/testExecutables.py]: I think the `exes` argument should be defaulted to `None` so it doesn't need to be specified explicitly.  Why create additional boilerplate?  In fact, can't the `create_executable_tests` call be folded in to the `class` definition somehow, e.g., in the `__init__` or through a metaclass?

Comment (timj) [inline tests/testExecutables.py]: This is out of scope for this ticket. It was covered in DM-5561. Please file a new ticket if you want the interface changed.

Comment (PaulPrice) [inline tests/testExecutables.py]: In fact, `exes` *does* default to `None`.  Get rid of `EXECUTABLES` then.

Comment (timj) [inline tests/testExecutables.py]: Yes it does. Thanks. The reason it is there is that this is present in the template version of this test file from the `utils` package. I did want the template to include the tuple so that people would understand that they have the option of including an explicit list in their code.

File: python/lsst/daf/base/baseLib.i (hunk 6 above)
Comment (timj) [inline python/lsst/daf/base/baseLib.i]: I've made the change and also fixed the logic with booleans. (bool inherits from int so I was converting bool to int by mistake).

File: python/lsst/daf/base/baseLib.i [hunk 13]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
 
+import numbers
+def _guessIntegerType(container, name, value):
+    """Given an existing container and name, determine the type
+    that should be used for the supplied value. The supplied value
+    is assumed to be a scalar.
+
+    On Python 3 all ints are LongLong but we need to be able to store them
+    in Int containers if that is what is being used (testing for truncation).
+    Int is assumed to mean 32bit integer (2147483647 to -2147483648).
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: It may well be on most systems these days, but are we going to get bitten by assuming an architecture?  Does `PropertySet` define it as a particular size, or just `int` ("at least 16 bits").

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: It seems to be `int`. @ktlim could we actually define this to use `int32_t` to make it more explicit?

File: python/lsst/daf/base/baseLib.i [hunk 14]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+    we pick Int if the value is less than maxsize.
+
+    Returns None if the value supplied is a bool or not an integral value.
+    """
+    useType = None
+    maxInt = 2147483647
+    minInt = -2147483648
+
+    # We do not want to convert bool to int so let the system work that
+    # out itself
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: How annoying that `isinstance(True, numbers.Integral)` is `True`!

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: My thoughts exactly when I saw the tests fail in `afw` last night.

File: python/lsst/daf/base/baseLib.i [hunk 15]
```diff
@@ -161,6 +161,48 @@ def _propertyContainerGet(container, name, asArray=False):
...
+            if value <= maxInt and value >= minInt:
+                useType = "Int"
+            else:
+                useType = "LongLong"
+        else:
+            if containerType == "Int":
+                if value <= maxInt and value >= minInt:
+                    useType = "Int"
+            elif containerType == "LongLong":
+                useType = "LongLong"
```
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: Does this mean that I'm going to get into trouble if I add 123 and then `maxInt+1`?

File: python/lsst/daf/base/baseLib.i (hunk 14 above)
Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: You had to wait for afw to fail before you knew there was a problem here?  Can you add a test that will catch the problem in this package?

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: I had to wait until `obs_sdss` for the integer overflow error to appear...

File: python/lsst/daf/base/baseLib.i (hunk 15 above)
Comment (timj) [inline python/lsst/daf/base/baseLib.i]: If you are using `Int` containers then yes. I'm not sure I understand you. Test case?

Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: What happens if:

    ps = PropertySet()
    ps.add("foo", 123)
    ps.add("foo", 2147483649)
    print(ps.get("foo"))
    print(ps.toString())

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: Without these patches:
```
Traceback (most recent call last):
  File "xxx.py", line 4, in <module>
    ps.add("foo", 2147483649)
  File "/Users/timj/work/lsstsw/stack/DarwinX86/daf_base/12.0-1-g22b9da8+3/python/lsst/daf/base/baseLib.py", line 4210, in _PS_addValue
    return _propertyContainerAdd(self, name, value, _PS_typeMenu)
  File "/Users/timj/work/lsstsw/stack/DarwinX86/daf_base/12.0-1-g22b9da8+3/python/lsst/daf/base/baseLib.py", line 4186, in _propertyContainerAdd
    return getattr(container, "add" + typeMenu[t])(name, value, *args)
  File "/Users/timj/work/lsstsw/stack/DarwinX86/daf_base/12.0-1-g22b9da8+3/python/lsst/daf/base/baseLib.py", line 3099, in addInt
    return _baseLib.PropertySet_addInt(self, *args)
OverflowError: Overflow error converting to int
```
with the new code (python2):
```
Traceback (most recent call last):
  File "xxx.py", line 4, in <module>
    ps.add("foo", 2147483649)
  File "/Users/timj/work/lsstsw/stack/DarwinX86/daf_base/tickets.DM-7028-ged7f34c3bf/python/lsst/daf/base/baseLib.py", line 4137, in _PS_addValue
    return _propertyContainerAdd(self, name, value, _PS_typeMenu)
  File "/Users/timj/work/lsstsw/stack/DarwinX86/daf_base/tickets.DM-7028-ged7f34c3bf/python/lsst/daf/base/baseLib.py", line 4112, in _propertyContainerAdd
    return getattr(container, "add" + addType)(name, value, *args)
  File "/Users/timj/work/lsstsw/stack/DarwinX86/daf_base/tickets.DM-7028-ged7f34c3bf/python/lsst/daf/base/baseLib.py", line 3099, in addInt
    return _baseLib.PropertySet_addInt(self, *args)
OverflowError: Overflow error converting to int
```

Comment (PaulPrice) [inline python/lsst/daf/base/baseLib.i]: OK, no change.  Let's leave it then.

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: Actually, there is a change of behavior on Python 3. On Python 2 "2147483649" is an `int` and not a `long` and so `addInt` is called which overflows. On Python 3 with the code above that falls through the type ends up as LongLong and when you call `addLongLong` on the `Int` item it triggers a type mismatch error. It seems that for compatibility I should check 64bit int range when the existing container type is Int and not the 32-bit range. This all seems to be a bit confused with Python `int` being a different size to PropertySet `Int`. I'll make the current behavior consistent with python 2 as it is indeed silly to shove a 64bit int into a 32-bit slot but I should let the later code tell you that.

Comment (timj) [inline python/lsst/daf/base/baseLib.i]: I've added tests and now ensure that OverflowError triggers consistently on Py2 and Py3. I think I'm ready to merge. Will do one last Jenkins test.
